cpp/src/arrow/record_batch.h

```diff
@@ -108,7 +108,7 @@ class ARROW_EXPORT RecordBatch {
   virtual std::shared_ptr<ArrayData> column_data(int i) const = 0;
 
   /// \brief Retrieve all arrays' internal data from the record batch.
-  virtual ArrayDataVector column_data() const = 0;
+  virtual const ArrayDataVector& column_data() const = 0;
```

Comment:
We would have to make it virtual and push it down into SimpleRecordBatch (which is ok because right now it makes O(cols) virtual calls internally anyways). That said, from a quick grep, I don't think the method is ever used. Table::columns has a few uses so maybe that's worth targeting. And it's probably better to be consistent where possible.

Reply:
I went ahead and converted Table::columns() and RecordBatch::columns().